### PR TITLE
Add support for xdg-output-v1

### PR DESCRIPTION
**Pull request by Dami-star**

### CODE REVIEW

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/server/qtquick/private/wquickxdgoutput.cpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+WQuickXdgOutputManager::WQuickXdgOutputManager(QObject *parent):
+    WQuickWaylandServerInterface(parent)
+  , WObject(*new WQuickXdgOutputManagerPrivate(this), nullptr)
+{
+
+}
+
+void WQuickXdgOutputManager::create(WOutputLayout *layout)
+{
+    wlr_xdg_output_manager_v1_create(server()->handle()->handle(), layout->handle());
```
[the remark this answers is not in the document]
Comment (Dami-star) [reply src/server/qtquick/private/wquickxdgoutput.cpp]: done

File: src/server/qtquick/private/wquickxdgoutput_p.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+class WAYLIB_SERVER_EXPORT WQuickXdgOutputManager : public WQuickWaylandServerInterface, public WObject
+{
+    Q_OBJECT
+    W_DECLARE_PRIVATE(WQuickXdgOutputManager)
+    QML_NAMED_ELEMENT(XdgOutputManager)
+
+public:
+    explicit WQuickXdgOutputManager(QObject *parent = nullptr);
+
+    void create(WOutputLayout *layout);
```
[the remark this answers is not in the document]
Comment (Dami-star) [reply src/server/qtquick/private/wquickxdgoutput_p.h]: done

File: src/server/qtquick/private/wquickxdgoutput.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    return d->layout;
+}
+
+void WQuickXdgOutputManager::create()
+{
+    W_D(WQuickXdgOutputManager);
+    WQuickWaylandServerInterface::create();
+    if(d->layout)
+        QWXdgOutputManagerV1::create(server()->handle(), d->layout);
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgoutput.cpp:64]: else qWarning

File: src/server/qtquick/private/wquickxdgoutput.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  , WObject(*new WQuickXdgOutputManagerPrivate(this), nullptr)
+{
+
+}
+
+void WQuickXdgOutputManager::setLayout(WOutputLayout *layout)
+{
+    W_D(WQuickXdgOutputManager);
+
+    if (d->layout == layout)
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgoutput.cpp:46]:
```
if (d->layout) {
  qmlWarning(this) << "Trying to set output layout for xdg output manager twice. Ignore this request";
  return;
}
```

**Review by wineee — CHANGES_REQUESTED** [review]

File: examples/tinywl/Main.qml [hunk 5]
```diff
@@ -182,6 +182,15 @@ Item {
                 })
             }
         }
+
+        XdgOutputManager {
+            id: xdgoutputmanager
+            layout: QmlHelper.layout
+
+            Component.onCompleted: {
+                console.info("XdgOutputManager")
```
Comment (wineee) [inline examples/tinywl/Main.qml:191]: remove

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/server/qtquick/private/wquickxdgoutput_p.h [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+public:
+    explicit WQuickXdgOutputManager(QObject *parent = nullptr);
+
+    void setLayout(WOutputLayout *layout);
+    WOutputLayout *layout() const;
+
+private:
+    void create() override;
+
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgoutput_p.h:31]:
```suggestion
```

File: src/server/qtquick/private/wquickxdgoutput_p.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
+// Copyright (C) 2024 Lu YaNing <luyaning@uniontech.org>.
+// SPDX-License-Identifier: Apache-2.0 OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+#pragma once
+
+#include <wglobal.h>
+#include <wquickwaylandserver.h>
+
+#include <QQmlEngine>
+
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgoutput_p.h:10]:
```suggestion
```

**Review by wineee — CHANGES_REQUESTED** [review]

File: src/server/qtquick/private/wquickxdgoutput.cpp (hunk 3 above)
Comment (Dami-star) [reply src/server/qtquick/private/wquickxdgoutput.cpp:66]: done

File: src/server/qtquick/private/wquickxdgoutput.cpp (hunk 4 above)
Comment (Dami-star) [reply src/server/qtquick/private/wquickxdgoutput.cpp]: done

File: examples/tinywl/Main.qml (hunk 5 above)
Comment (Dami-star) [reply examples/tinywl/Main.qml]: done

File: src/server/qtquick/private/wquickxdgoutput_p.h (hunk 6 above)
Comment (Dami-star) [reply src/server/qtquick/private/wquickxdgoutput_p.h]: done

File: src/server/qtquick/private/wquickxdgoutput_p.h (hunk 7 above)
Comment (Dami-star) [reply src/server/qtquick/private/wquickxdgoutput_p.h:9]: done

File: examples/tinywl/Main.qml [hunk 8]
```diff
@@ -182,6 +182,11 @@ Item {
                 })
             }
         }
+
+        XdgOutputManager {
+            id: xdgoutputmanager
```
Comment (wineee) [inline examples/tinywl/Main.qml:187]:
```suggestion
            id: xdgOutputManager
```

File: src/server/qtquick/private/wquickxdgoutput.cpp [hunk 9]
```diff
@@ -0,0 +1,71 @@
+// Copyright (C) 2024 Lu YaNing <luyaning@uniontech.org>.
+// SPDX-License-Identifier: Apache-2.0 OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+#include "wquickxdgoutput_p.h"
+#include "wsurface.h"
+#include "wxdgsurface.h"
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgoutput.cpp:6]: wsurface and wxdgsurface should unused?

**Review by wineee — CHANGES_REQUESTED** [review]

File: examples/tinywl/Main.qml (hunk 8 above)
Comment (Dami-star) [reply examples/tinywl/Main.qml]: done

File: src/server/qtquick/private/wquickxdgoutput.cpp (hunk 9 above)
Comment (Dami-star) [reply src/server/qtquick/private/wquickxdgoutput.cpp]: done

File: src/server/qtquick/private/wquickxdgoutput.cpp [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+void WQuickXdgOutputManager::create()
+{
+    W_D(WQuickXdgOutputManager);
+    WQuickWaylandServerInterface::create();
+    if(d->layout)
+        QWXdgOutputManagerV1::create(server()->handle(), d->layout);
+    else
+        qWarning() << "Need to set output layout !";
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgoutput.cpp:66]:
```suggestion
        qWarning() << "Output layout not set, xdg output manager will never be created!";
```

**Review by wineee — CHANGES_REQUESTED** [review]

**Comment by wineee** [comment]
also need rebase to have https://github.com/vioken/waylib/pull/330

File: src/server/qtquick/private/wquickxdgoutput.cpp (hunk 10 above)
Comment (Dami-star) [reply src/server/qtquick/private/wquickxdgoutput.cpp]: done

**Review by wineee — APPROVED** [review]
LGTM, @zccrs

File: src/server/qtquick/private/wquickxdgoutput.cpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+    }
+
+    W_DECLARE_PUBLIC(WQuickXdgOutputManager)
+
+    WOutputLayout *layout = nullptr;
+};
+
+WQuickXdgOutputManager::WQuickXdgOutputManager(QObject *parent):
+    WQuickWaylandServerInterface(parent)
```
[the remark this answers is not in the document]
Comment (Dami-star) [reply src/server/qtquick/private/wquickxdgoutput.cpp]: done

File: src/server/qtquick/private/wquickxdgoutput.cpp [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    W_DC(WQuickXdgOutputManager);
+
+    return d->layout;
+}
+
+void WQuickXdgOutputManager::create()
+{
+    W_D(WQuickXdgOutputManager);
+    WQuickWaylandServerInterface::create();
+    if(d->layout)
```
[the remark this answers is not in the document]
Comment (Dami-star) [reply src/server/qtquick/private/wquickxdgoutput.cpp]: done

**Review by zccrs — APPROVED** [review]

**Comment by wineee** [comment]
?? @Dami-star  https://github.com/vioken/waylib/compare/4c010bec24731af50a7d175bdb11c61fa0696874..ee0fc5df75c108e9e04e706a493efedf11631c2d

**Review by zccrs — APPROVED** [review]

**Pull request merged**